File: easy/796-rotate-string/solution.py
class Solution:
    def rotateString(self, s: str, goal: str) -> bool:
        """
        796. Rotate String
        Time: O(n)
        Space: O(n)
        """
        if len(s) != len(goal):
            return False

        pattern = goal
        text = s + s
        lps = self._build_lps(pattern)
        matched = 0

        for char in text:
            while matched > 0 and char != pattern[matched]:
                matched = lps[matched - 1]

            if char == pattern[matched]:
                matched += 1

            if matched == len(pattern):
                return True

        return False

    def _build_lps(self, pattern: str) -> list[int]:
        lps = [0] * len(pattern)
        length = 0

        for i in range(1, len(pattern)):
            while length > 0 and pattern[i] != pattern[length]:
                length = lps[length - 1]

            if pattern[i] == pattern[length]:
                length += 1
                lps[i] = length

        return lps

File: easy/796-rotate-string/solution.java
/**
 * 796. Rotate String
 * Time: O(n)
 * Space: O(n)
 */
class Solution {
    public boolean rotateString(String s, String goal) {
        if (s.length() != goal.length()) {
            return false;
        }

        String text = s + s;
        int[] lps = buildLps(goal);
        int matched = 0;

        for (int i = 0; i < text.length(); i++) {
            while (matched > 0 && text.charAt(i) != goal.charAt(matched)) {
                matched = lps[matched - 1];
            }

            if (text.charAt(i) == goal.charAt(matched)) {
                matched++;
            }

            if (matched == goal.length()) {
                return true;
            }
        }

        return false;
    }

    private int[] buildLps(String pattern) {
        int[] lps = new int[pattern.length()];
        int length = 0;

        for (int i = 1; i < pattern.length(); i++) {
            while (length > 0 && pattern.charAt(i) != pattern.charAt(length)) {
                length = lps[length - 1];
            }

            if (pattern.charAt(i) == pattern.charAt(length)) {
                length++;
                lps[i] = length;
            }
        }

        return lps;
    }
}
